### DISCUSSION TITLE
Wrong LazyGhost initialization on cloned entity

### DISCUSSION THREAD

**Discussion by Erwui** [symfony/symfony#50874] [Q&A] [answered]
Hello,

With the new version of proxy doctrine using the lazyGhost object I'm encountering strange behavior during initialization. The cloned entity is "loaded" with data from another entry that has nothing to do with it. It mutates completely at random. This produces a "duplicate entry" sql error, as doctrine is trying to insert data that already exists.
When I set "enable_lazy_ghost_objects" to false, everything works fine.

![Capture-2023-07-04-150325](https://github.com/symfony/symfony/assets/26304231/d3f32e95-46aa-4e1a-8c42-60ca2b6bf387)

The code below. I don't do anything except set a collection to initialise the entity between the two dumps

```php
dump(initial: $analysisResult, clone: $analysisResultClone);

$analysisResultClone->setActivityPhases(new ArrayCollection());

dump(cloneInitializedFull: $analysisResultClone);
```

Thanks for your help

**Comment by stof** [comment]
@nicolas-grekas do you have any idea ?

**Reply by nicolas-grekas** [reply]
Nope. But I'd be happy to have a look at a bug report (on https://github.com/doctrine/orm I suppose), with a reproducer please!

**Reply by MatTheCat** [reply]
The issue comes from the fact the proxy’s identifier is `null`.

`ClassMetadataInfo::getIdentifierValues` then returns `[]` which is used as a filter by `BasicEntityPersister` which then loads the whole table to only return the first corresponding entity.

Maybe the initializer should throw when a proxy has no identifier?

**Reply by Erwui** [reply]
You can find the reproducer here: https://github.com/Erwui/lazy_ghost_bug
I open the issue doctrine about this.

**Reply by MatTheCat** [reply]
~Okay so the actual bug is on your side: you shouldn’t unset proxies’ identifiers.~

EDIT: see https://github.com/symfony/symfony/discussions/50874#discussioncomment-6360556

**Reply by Erwui** [reply]
The LazyGhost object has changed the behavior, this worked correctly on the previous version or when I disable lazyGhost object.
If I can't unset the id, what am I supposed to do?
For a simple use case, for example, I want to duplicate a blog post. I have to create a new one and define the properties from the original manually, whereas the clone function in php is designed for this ?

**Reply by MatTheCat** [reply]
Okay it seems the main difference is that when `enable_lazy_ghost_objects` is `true`, `$barClone` will be initialized in `Bar::__clone` when `$this->bazs = new ArrayCollection();` is called, thus falling in [the case I mentioned](https://github.com/symfony/symfony/discussions/50874#discussioncomment-6355418) since the identifier is unset just before.

When `enable_lazy_ghost_objects` is `false`, Doctrine will first call its cloner which will initialize `$barClone`, **then** `Bar::__clone` will be called.

**Reply by MatTheCat** [reply, marked as answer]
I have test cases and fixes. Will open PRs today.

EDIT: see https://github.com/symfony/symfony/pull/50899 and https://github.com/doctrine/orm/pull/10818

**Reply by Erwui** [reply]
Thank you @MatTheCat for the fix ! Glad I could help you.
